### PR TITLE
chore: Make iterators working again with convert

### PR DESCRIPTION
**Pull request by mutahhir**
### Description

Due to the recent TF expression AST work, we have had to re-adapt the recent iterator work from https://github.com/hashicorp/terraform-cdk/pull/2732 to work with the new AST conversion.

### Checklist

- [ ] I have updated the PR title to match [CDKTF's style guide](https://github.com/hashicorp/terraform-cdk/blob/main/CONTRIBUTING.md#pull-requests-1)
- [ ] I have run the linter on my code locally
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation if applicable
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works if applicable
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

**Comment by team-tf-cdk** [comment]
This PR exceeds the recommended size of 1000 lines. Please make sure you are NOT addressing multiple issues with one PR. Note this PR might be rejected due to its size.

File: packages/@cdktf/hcl2cdk/test/expressionToTs.test.ts [hunk 1]
```diff
@@ -148,7 +200,11 @@ describe("expressionToTs", () => {
...
     const expression = `"simple-\${local.foo}"`;
     const scope = getScope({ locals: ["foo"] });
-    const result = await convertTerraformExpressionToTs(expression, scope, "");
+    const result = await convertTerraformExpressionToTs(
+      expression,
+      scope,
+      getType
+    );
     expect(code(result)).toMatchInlineSnapshot(
       `"\\"simple-\${\\" + foo + \\"}\\""`
```
Comment (ansgarm) [inline packages/@cdktf/hcl2cdk/test/expressionToTs.test.ts:209]: given that this expects `foo` to be the variable name for that local – wouldn't this collide in the case of HCL having both a `var.foo` and a `local.foo`?
Or am I missing something?

Comment (ansgarm) [reply packages/@cdktf/hcl2cdk/test/expressionToTs.test.ts:209]: (this is probably not related to this PR, but rather an existing shortcoming – if true)

File: packages/@cdktf/hcl2cdk/test/expressionToTs.test.ts [hunk 2]
```diff
@@ -198,37 +270,53 @@ describe("expressionToTs", () => {
...
     const scope = getScope({ resources: ["aws_s3_bucket.examplebucket"] });
-    const result = await convertTerraformExpressionToTs(expression, scope, "");
+    const result = await convertTerraformExpressionToTs(
+      expression,
+      scope,
+      getType
+    );
     expect(code(result)).toMatchInlineSnapshot(
-      `"cdktf.Fn.toset(\\"\${\\" + awsS3BucketExamplebucket.fqn + \\"}.*\\")"`
+      `"cdktf.Token.asString(cdktf.Fn.toset(\\"\${\\" + awsS3BucketExamplebucket.fqn + \\"}.*\\"))"`
```
Comment (ansgarm) [inline packages/@cdktf/hcl2cdk/test/expressionToTs.test.ts:279]: For this test case it doesn't really make sense to have a `getType` function that returns string 😄

Comment (ansgarm) [reply packages/@cdktf/hcl2cdk/test/expressionToTs.test.ts:279]: (this is also true for some other test cases)

**Review by ansgarm — APPROVED** [review]
Nice work – left some comments, but iirc none of these would block a merge from my point of view 😄

File: packages/@cdktf/hcl2cdk/lib/types.ts [hunk 3]
```diff
@@ -24,4 +24,5 @@ export type ProgramScope = {
 export type ResourceScope = ProgramScope & {
   forEachIteratorName?: string; // set if a for_each is used in this resource
   withinOverrideExpression?: boolean; // set if we are currently within an override expression
+  scopedVariables?: Record<string, string>; // set if we are currently within an dynamic block expression
```
[the remark this answers is not in the document]
Comment (mutahhir) [reply packages/@cdktf/hcl2cdk/lib/types.ts:27]: So, 'dynamic-block' is kind of a constant we use to represent current and ancestor dynamic block names, but this was originally designed to be used with iterators.

File: packages/@cdktf/hcl2cdk/test/expressionToTs.test.ts (hunk 2 above)
Comment (mutahhir) [reply packages/@cdktf/hcl2cdk/test/expressionToTs.test.ts:279]: :) indeed. This is a bit noisy

File: packages/@cdktf/hcl2cdk/test/__snapshots__/iteration.test.ts.snap [hunk 4]
```diff
@@ -33,23 +33,35 @@ class MyConvertedCode extends cdktf.TerraformStack {
...
       \\"example_3\\",
       {
         allowOverwrite: true,
-        name: \\"\${each.value.name}\\",
-        records: [\\"\${each.value.record}\\"],
+        name: cdktf.Token.asString(
+          cdktf.propertyAccess(awsRoute53RecordExampleForEachIterator.value, [
+            \\"name\\",
+          ])
+        ),
```
[the remark this answers is not in the document]
Comment (mutahhir) [reply packages/@cdktf/hcl2cdk/test/__snapshots__/iteration.test.ts.snap:51]: Indeed, we could do more here.

File: packages/@cdktf/hcl2cdk/lib/generation.ts [hunk 5]
```diff
@@ -62,133 +62,18 @@ function getReference(graph: DirectedGraph, id: string) {
...
-      return await convertTerraformExpressionToTs(`"${item}"`, scope, path);
-
-    // return coerceType(
-    //   scope,
-    //   ast,
-    //   findTypeOfReference(scope, ast, references),
-    //   getDesiredType(scope, path)
-    // );
+      return await convertTerraformExpressionToTs(`"${item}"`, scope, () =>
+        getDesiredType(scope, path)
```
[the remark this answers is not in the document]
Comment (mutahhir) [reply packages/@cdktf/hcl2cdk/lib/generation.ts:74]: I think we could revisit it, your comments make sense. We hashed it out in a hurry during a pairing session, so I'm sure there's a way to improve.

**Pull request merged**